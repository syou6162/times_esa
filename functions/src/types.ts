/**
 * Firebase Functionsのリクエスト/レスポンス型定義
 */

import { type DailyReportCategory } from './dateUtils';

// 設定型定義
export type EsaConfig = {
  teamName: string;
  accessToken: string;
}

// リクエスト型定義
export type SubmitTextRequest = {
  category: DailyReportCategory;
  tags: string[];
  title: string;
  text: string;
}

export type DailyReportRequest = {
  category: DailyReportCategory;
}

export type RecentDailyReportsRequest = {
  days?: number; // 過去何日分を取得するか（デフォルト: 10）
}

// エラーレスポンス型定義
export type EsaErrorResponse = {
  error: string;
  message: string;
}