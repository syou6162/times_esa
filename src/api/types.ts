import type { HttpsCallableResult } from 'firebase/functions';

// リクエスト/レスポンスの型定義
export type DailyReportRequest = {
  category: string;
}

export type DailyReportResponse = {
  updatedAt: string;
  url: string;
  bodyMd: string;
  bodyHtml: string;
  tags: string[];
  name: string;
  category: string;
}

export type TagListRequest = {}

export type TagListResponse = {
  tags: Array<{
    name: string;
    postsCount: number;
  }>;
  totalCount: number;
}

export type SubmitTextToEsaRequest = {
  category: string;
  tags: string[];
  title: string;
  text: string;
}

export type SubmitTextToEsaResponse = {
  updatedAt: string;
  url: string;
  bodyMd: string;
  bodyHtml: string;
  tags: string[];
  name: string;
  category: string;
}

export type RecentDailyReportsRequest = {}

export type RecentDailyReportsResponse = {
  reports: Array<{
    date: string;
    url: string;
    posts_count: number;
  }>;
}

// APIクライアントのインターフェース
export interface ApiClient {
  getDailyReport(category: string): Promise<HttpsCallableResult<DailyReportResponse>>;
  getTagList(): Promise<HttpsCallableResult<TagListResponse>>;
  submitTextToEsa(
    category: string,
    tags: string[],
    title: string,
    text: string
  ): Promise<HttpsCallableResult<SubmitTextToEsaResponse>>;
  getRecentDailyReports(): Promise<HttpsCallableResult<RecentDailyReportsResponse>>;
}