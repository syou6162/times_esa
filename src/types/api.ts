/**
 * API関連の型定義
 * 
 * APIのリクエスト・レスポンスに関する型を定義
 * 基本的にAPIの仕様に合わせてsnake_caseを使用
 */

// 日報取得API
export type DailyReportRequestType = {
  category: string;
};

export type DailyReportResponseType = {
  updated_at: string;
  url: string;
  body_md: string;
  body_html: string;
  tags: string[];
  name: string;
  category: string;
};

// タグ一覧取得API
export type TagListRequestType = Record<string, never>;

export type TagListResponseType = {
  tags: Array<{
    name: string;
    posts_count: number;
  }>;
};

// 最近の日報一覧取得API
export type RecentDailyReportsRequest = {
  days?: number; // 過去何日分を取得するか（デフォルト: 10）
};

export type RecentDailyReportsResponse = {
  reports: Array<{
    date: string;
    title: string;
    posts_count: number;
    tags: string[];
    formatted_date: string;
    url?: string;
  }>;
  total_count: number;
};

// ESA投稿API
export type SubmitTextToEsaRequestType = {
  category: string;
  tags: string[];
  title: string;
  text: string;
};

export type SubmitTextToEsaResponseType = {
  updated_at: string;
  url: string;
  body_md: string;
  body_html: string;
  tags: string[];
  name: string;
  category: string;
};