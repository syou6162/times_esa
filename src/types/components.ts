// コンポーネントのProps型定義をまとめたファイル

import { Auth } from 'firebase/auth';
import { AutocompleteProps } from '@mui/material/Autocomplete';
import { GoogleUser, InputChangeHandler, TextAreaChangeHandler } from './index';

// Body Component Props
export type BodyProps = {
  hasUserLanded: boolean;
  isSignedIn: boolean;
  user: GoogleUser;
  firebaseAuth: Auth | null;
}

// DailyReport Component Props
export type DailyReportProps = {
  fetching: boolean;
  fetchErrorMessage: string;
  esaText: string;
  esaHtml: string;
  reloadDailyReport: () => void;
}

// DailyReport内部のコンポーネントProps
export type DailyReportHtmlProps = {
  esaHtml: string;
}

export type DailyReportTextProps = {
  esaText: string;
}

export type DailyReportShareProps = {
  esaText: string;
}

// ShareButton Props
export type TweetButtonProps = {
  text: string;
}

export type CopyButtonProps = {
  text: string;
}

// EsaSubmitForm Props
export type EsaSubmitFormProps = {
  category: string;
  title: string;
  tags: string[];
  tagCandidates: string[];
  fetching: boolean;
  onSubmit: (
    category: string,
    title: string,
    markdown: string,
    html: string,
    tags: string[]
  ) => void;
}

// EsaTitleField Props
export type EsaTitleFieldProps = {
  fetching: boolean;
  sending: boolean;
  title: string;
  onChange: InputChangeHandler;
}

// EsaTextField Props
export type EsaTextFieldProps = {
  sending: boolean;
  text: string;
  onChange: TextAreaChangeHandler;
}

// Autocomplete用の型エイリアス
type AutocompleteChangeHandler<T> = AutocompleteProps<T, true, false, true>['onChange'];

// EsaTagsField Props
export type EsaTagsFieldProps = {
  fetching: boolean;
  sending: boolean;
  tags: string[];
  tagCandidates: string[];
  onChange: AutocompleteChangeHandler<string>;
}

// TimesEsa Props
export type TimesEsaProps = {
  canFetchCloudFunctionEndpoints: boolean;
}

// DailyReportsList Props
export type DailyReportsListProps = {
  selectedDate?: string;
  onDateSelect: (date: string) => void;
}

// SignInDialog Props
export type SignInDialogProps = {
  firebaseAuth: Auth | null;
}

// SignOutButton Props
export type SignOutButtonProps = {
  firebaseAuth: Auth | null;
}

// StyledFirebaseAuth Props
export type StyledFirebaseAuthProps = {
  firebaseAuth: Auth | null;
}

// Footer Props
export type FooterProps = {
  isSignedIn: boolean;
  user: GoogleUser;
  firebaseAuth: Auth | null;
}