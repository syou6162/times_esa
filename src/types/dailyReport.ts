/**
 * 日報関連の型定義
 */

export type Tag = {
  name: string;
  posts_count: number; // eslint-disable-line camelcase
};

export type DailyReportSummary = {
  date: string; // yyyy-MM-dd形式
  title: string;
  postsCount: number;
  tags: string[];
  formattedDate: string; // 日本語日付（例: "6月19日(水)"）
  url?: string; // esaのURL
};

export type DailyReportContent = {
  updated_at: string;
  url: string;
  body_md: string;
  body_html: string;
  tags: string[];
  name: string;
  category: string;
};

export type DailyReportRequestType = {
  category: string;
};

export type DailyReportResponseType = DailyReportContent;

export type TagListRequestType = Record<string, never>;

export type TagListResponseType = {
  tags: Tag[];
};

export type RecentDailyReportsRequest = {
  days?: number; // 過去何日分を取得するか（デフォルト: 10）
};

export type RecentDailyReportsResponse = {
  reports: DailyReportSummary[];
  total_count: number;
};

/**
 * 日報の表示モード
 */
export type ReportViewMode = 'today' | 'past';

/**
 * 日報選択の状態
 */
export type SelectedDateState = 'today' | string; // 'today' または yyyy-MM-dd形式