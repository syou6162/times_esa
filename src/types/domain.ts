/**
 * ドメインモデルの型定義
 * 
 * アプリケーション内で使用するドメインモデルの型を定義
 * 基本的にcamelCaseを使用
 */

// タグ情報
export type Tag = {
  name: string;
  posts_count: number; // eslint-disable-line camelcase
};

// 日報サマリー（一覧表示用）
export type DailyReportSummary = {
  date: string; // yyyy-MM-dd形式
  title: string;
  postsCount: number;
  tags: string[];
  formattedDate: string; // 日本語日付（例: "6月19日(水)"）
  url?: string; // esaのURL
};

// 日報コンテンツ（API responseの形式）
export type DailyReportContent = {
  updated_at: string;
  url: string;
  body_md: string;
  body_html: string;
  tags: string[];
  name: string;
  category: string;
};

// 日報データ（アプリ内部で使用する形式、camelCase）
export type DailyReportData = {
  updatedAt: string;
  url: string;
  text: string;      // body_md → text
  html: string;      // body_html → html
  tags: string[];
  title: string;     // name → title
  category: string;
};

// ユーザー情報
export type GoogleUser = {
  email: string;
  displayName: string;
  photoURL: string;
};

// 日報の表示モード
export type ReportViewMode = 'today' | 'past';

// 日報選択の状態
export type SelectedDateState = 'today' | string; // 'today' または yyyy-MM-dd形式