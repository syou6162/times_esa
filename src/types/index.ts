// 共通の型定義をまとめたファイル

// ユーザー情報の型定義
export type GoogleUser = {
  email: string;
  displayName: string;
  photoURL: string;
}

// タグの型定義（バックエンドのEsaTagCamelCaseと同じ構造）
export type Tag = {
  name: string;
  postsCount: number;
}

// 日報表示タイプのenum
export enum DailyReportType {
  HTML = 'HTML',
  TEXT = 'TEXT',
  SHARE = 'SHARE',
}

// Form要素のイベントハンドラー型
export type InputChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => void;
export type TextAreaChangeHandler = (e: React.ChangeEvent<HTMLTextAreaElement>) => void;
