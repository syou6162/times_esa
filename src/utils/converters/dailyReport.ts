/**
 * 日報データの変換関数
 * 
 * ESA APIのレスポンス（snake_case）をアプリ内部のモデル（camelCase）に変換
 */

import { 
  DailyReportResponseType, 
  TagListResponseType,
  RecentDailyReportsResponse 
} from '../../types/api';
import { 
  DailyReportData, 
  Tag,
  DailyReportSummary 
} from '../../types/domain';

/**
 * 日報レスポンスをアプリ内部のデータ形式に変換
 */
export const convertDailyReportResponse = (response: DailyReportResponseType): DailyReportData => {
  return {
    updatedAt: response.updated_at,
    url: response.url,
    text: response.body_md,
    html: response.body_html,
    tags: response.tags,
    title: response.name,
    category: response.category,
  };
};

/**
 * タグレスポンスをアプリ内部のデータ形式に変換
 * 
 * 現状はそのまま返すだけだが、将来的な変更に備えて変換関数を用意
 */
export const convertTagListResponse = (response: TagListResponseType): Tag[] => {
  return response.tags;
};

/**
 * 最近の日報一覧レスポンスをアプリ内部のデータ形式に変換
 */
export const convertRecentReportsResponse = (response: RecentDailyReportsResponse): DailyReportSummary[] => {
  return response.reports.map(report => ({
    date: report.date,
    title: report.title,
    postsCount: report.posts_count,  // snake_case → camelCase
    tags: report.tags,
    formattedDate: report.formatted_date,  // snake_case → camelCase
    url: report.url,
  }));
};