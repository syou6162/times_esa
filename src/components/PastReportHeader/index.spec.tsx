// @vitest-environment jsdom
import { render } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import { PastReportHeader, PastReportHeaderProps } from '.';
import { DailyReportSummary } from '../../types/dailyReport';

const mockReport: DailyReportSummary = {
  date: '2024-06-19',
  title: '開発、会議、レビュー',
  postsCount: 5,
  tags: ['開発', 'times_esa'],
  formattedDate: '6月19日(水)'
};

describe('PastReportHeaderが正しく表示される', () => {
  it('読み取り専用モードで表示される', () => {
    const props: PastReportHeaderProps = {
      report: mockReport,
      isReadOnly: true,
    };

    const renderResult = render(<PastReportHeader {...props} />);

    expect(renderResult.asFragment()).toMatchSnapshot();
    expect(renderResult.getByText('開発、会議、レビュー')).toBeTruthy();
    expect(renderResult.getByText('6月19日(水)の日報')).toBeTruthy();
    expect(renderResult.getByText('開発')).toBeTruthy();
    expect(renderResult.getByText('times_esa')).toBeTruthy();
    expect(renderResult.getByText('5個のつぶやき')).toBeTruthy();
  });

  it('編集可能モードで表示される', () => {
    const props: PastReportHeaderProps = {
      report: mockReport,
      isReadOnly: false,
    };

    const renderResult = render(<PastReportHeader {...props} />);

    expect(renderResult.asFragment()).toMatchSnapshot();
    expect(renderResult.getByText('6月19日(水)の日報')).toBeTruthy();
    expect(renderResult.getByText('5個のつぶやき')).toBeTruthy();
    // isReadOnlyがfalseの場合でも、つぶやき数が表示されることを確認
    expect(renderResult.queryByText('過去の日報は編集できません')).toBeNull();
  });

  it('レポートURLがある場合リンクとして表示される', () => {
    const reportWithoutTags: DailyReportSummary = {
      ...mockReport,
      tags: [],
      url: 'https://example.esa.io/posts/123',
    };
    const props: PastReportHeaderProps = {
      report: reportWithoutTags,
      isReadOnly: true,
    };

    const renderResult = render(<PastReportHeader {...props} />);
    
    const link = renderResult.getByRole('link', { name: '6月19日(水)の日報' });
    expect(link).toBeTruthy();
    expect(link.getAttribute('href')).toBe('https://example.esa.io/posts/123');
    expect(link.getAttribute('target')).toBe('_blank');
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('レポートURLがない場合テキストとして表示される', () => {
    const reportWithZeroPosts: DailyReportSummary = {
      ...mockReport,
      postsCount: 5,
    };
    const props: PastReportHeaderProps = {
      report: reportWithZeroPosts,
      isReadOnly: true,
    };

    const renderResult = render(<PastReportHeader {...props} />);

    // URLがない場合、リンクではなくテキストとして表示される
    expect(renderResult.queryByRole('link')).toBeNull();
    expect(renderResult.getByText('6月19日(水)の日報')).toBeTruthy();
  });
});