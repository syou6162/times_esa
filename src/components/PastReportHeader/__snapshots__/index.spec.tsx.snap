// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PastReportHeaderが正しく表示される > タグが空の場合も正しく表示される 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i3pbo"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1xbcbn6-MuiTypography-root"
    >
      6月19日(水)の日報（読み取り専用）
    </h5>
    <h6
      class="MuiTypography-root MuiTypography-h6 css-x3ur3e-MuiTypography-root"
    >
      開発、会議、レビュー
    </h6>
    <div
      class="MuiBox-root css-1ppihqm"
    />
    <p
      class="MuiTypography-root MuiTypography-body2 css-jyouah-MuiTypography-root"
    >
      過去の日報は編集できません・5個のつぶやき
    </p>
  </div>
</DocumentFragment>
`;

exports[`PastReportHeaderが正しく表示される > 編集可能モードで表示される 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i3pbo"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1xbcbn6-MuiTypography-root"
    >
      6月19日(水)の日報
    </h5>
    <h6
      class="MuiTypography-root MuiTypography-h6 css-x3ur3e-MuiTypography-root"
    >
      開発、会議、レビュー
    </h6>
    <div
      class="MuiBox-root css-1ppihqm"
    >
      <div
        class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-outlinedDefault css-2mpvrf-MuiChip-root"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-16cgrcw-MuiChip-label"
        >
          開発
        </span>
      </div>
      <div
        class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-outlinedDefault css-2mpvrf-MuiChip-root"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-16cgrcw-MuiChip-label"
        >
          times_esa
        </span>
      </div>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body2 css-jyouah-MuiTypography-root"
    >
      5個のつぶやき
    </p>
  </div>
</DocumentFragment>
`;

exports[`PastReportHeaderが正しく表示される > 読み取り専用モードで表示される 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i3pbo"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1xbcbn6-MuiTypography-root"
    >
      6月19日(水)の日報（読み取り専用）
    </h5>
    <h6
      class="MuiTypography-root MuiTypography-h6 css-x3ur3e-MuiTypography-root"
    >
      開発、会議、レビュー
    </h6>
    <div
      class="MuiBox-root css-1ppihqm"
    >
      <div
        class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-outlinedDefault css-2mpvrf-MuiChip-root"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-16cgrcw-MuiChip-label"
        >
          開発
        </span>
      </div>
      <div
        class="MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-outlinedDefault css-2mpvrf-MuiChip-root"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-16cgrcw-MuiChip-label"
        >
          times_esa
        </span>
      </div>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body2 css-jyouah-MuiTypography-root"
    >
      過去の日報は編集できません・5個のつぶやき
    </p>
  </div>
</DocumentFragment>
`;
